examples/credential-registry-storage-contract/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+        self.encrypted_credential
+            .serial::<Vec<_>>(bytes)
+            .map_err(|_| CustomContractError::SerializationError)?;
+        self.metadata.serial(bytes).map_err(|_| CustomContractError::SerializationError)?;
+        self.timestamp.serial(bytes).map_err(|_| CustomContractError::SerializationError)?;
+        Ok(())
+    }
+}
+
+/// The parameter type for the contract function `serializationHelper`.
+#[derive(Serialize, SchemaType)]
+pub struct SerializationHelperParam {
+    /// The contract_address that the signature is intended for.
+    contract_address:     ContractAddress,
+    /// The serialized encrypted_credential.
+    #[concordium(size_length = 2)]
+    encrypted_credential: Vec<u8>,
+    /// Metadata associated with the credential.
+    metadata:             Metadata,
+    /// A timestamp to make signatures expire.
+    timestamp:            Timestamp,
+}
+
+/// Helper function that can be invoked at the front-end to serialize
+/// the message to be signed by the wallet.
+#[receive(
+    contract = "credential-registry-storage",
+    name = "serializationHelper",
+    parameter = "SerializationHelperParam"
+)]
+fn contract_serialization_helper<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    Ok(())
+}
+
+/// Store credential data.
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - The signature was intended for a different contract.
+/// - The signature was intended for a different `entry_point`.
+/// - The signature is expired.
+/// - The signature can not be validated.
+/// - A credential is already registered for the given `public_key`.
+/// - Fails to log event.
+#[receive(
+    contract = "credential-registry-storage",
+    name = "store",
+    parameter = "StoreParam",
+    error = "CustomContractError",
+    crypto_primitives,
+    enable_logger,
+    mutable
+)]
+fn store<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ContractResult<()> {
+    // Parse the parameter.
+    let param: StoreParam = ctx.parameter_cursor().get()?;
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(param.contract_address, ctx.self_address(), CustomContractError::WrongContract);
+
+    // Check signature is not expired.
+    ensure!(param.timestamp > ctx.metadata().slot_time(), CustomContractError::Expired);
+
+    // Perepare message bytes as it is signed by the wallet.
+    // Note that the message is prepended by a domain separation string.
+    let mut message: Vec<u8> = SIGNARUTE_DOMAIN.as_bytes().to_vec();
+    param.message_bytes(&mut message)?;
+
+    // Calculate the message hash.
+    let message_hash = crypto_primitives.hash_sha2_256(&message).0;
```

Comment:
We don't really need to hash it since we are designing a new protocol anyhow. The signature scheme hashes the message already anyhow.

Reply:
That should make it cheaper as well.